pkg/kafka/correlation_cache.go

```diff
@@ -0,0 +1,224 @@
+
+	// nextSequenceNumber is the next sequence number to be used as
+	// correlation ID
+	nextSequenceNumber CorrelationID
+
+	// stopGc is closed when the garbage collector must exit
+	stopGc chan struct{}
+}
+
+// NewCorrelationCache returns a new correlation cache
+func NewCorrelationCache() *CorrelationCache {
+	cc := &CorrelationCache{
+		cache:              requestsCache{},
+		nextSequenceNumber: 1,
+		stopGc:             make(chan struct{}),
+	}
+
+	go cc.garbageCollector()
+
+	return cc
+}
+
+// DeleteCache releases the cache and stops the garbage collector. This
+// function must be called when the cache is no longer required, otherwise go
+// routines are leaked.
+func (cc *CorrelationCache) DeleteCache() {
+	close(cc.stopGc)
+}
+
+// HandleRequest must be called when a request is forwarded to the broker, will
+// keep track of the request and rewrite the correlation ID inside of the
+// request to a sequence number. This sequence number is guaranteed to be
+// unique within the connection covered by the cache.
+func (cc *CorrelationCache) HandleRequest(req *RequestMessage, finishFunc FinishFunc) {
+	cc.mutex.Lock()
+	defer cc.mutex.Unlock()
+
+	// save the original correlation ID
+	origCorrelationID := req.GetCorrelationID()
+
+	// Use a sequence number to generate a correlation ID that is
+	// guaranteed to be unique
+	newCorrelationID := cc.nextSequenceNumber
+	cc.nextSequenceNumber++
+
+	// Overwrite the correlation ID in the request to allow correlating the
+	// response later on. The original correlation ID will be restored when
+	// forwarding the response
+	req.SetCorrelationID(newCorrelationID)
+
+	if _, ok := cc.cache[newCorrelationID]; ok {
+		log.Warning("BUG: Overwriting Kafka request message in correlation cache")
+	}
+
+	cc.cache[newCorrelationID] = &correlationEntry{
+		request:           req,
+		created:           time.Now(),
+		origCorrelationID: origCorrelationID,
+		finishFunc:        finishFunc,
+	}
+}
+
+// Delete removes a request from the correlation cache. After deletion, the
+// request will no longer be correlated with responses. Note: Calling Delete()
+// will *not* invoke the FinishFunc of the request
+func (cc *CorrelationCache) Delete(req *RequestMessage) {
+	cc.mutex.Lock()
+	defer cc.mutex.Unlock()
+
+	correlationID := req.GetCorrelationID()
+	if correlationID != 0 {
+		delete(cc.cache, correlationID)
+	}
+}
+
+// correlate returns the request message with the matching correlation ID
+func (cc *CorrelationCache) correlate(id CorrelationID) *correlationEntry {
+	cc.mutex.RLock()
+	defer cc.mutex.RUnlock()
+
+	entry, _ := cc.cache[id]
+	return entry
+}
+
+// CorrelateResponse extracts the correlation ID from the response message,
+// correlates the corresponding request, restores the original correlation ID
+// in the response and returns the original request
+func (cc *CorrelationCache) CorrelateResponse(res *ResponseMessage) *RequestMessage {
+	entry := cc.correlate(res.GetCorrelationID())
```

Comment:
I am wondering if there can be a possible race here:
i.e between `correlate` which correlates an entry and `delete` which is suppose to delete that entry?

Scenario:
Request 1: Correlation ID : 5 
correlate return 5 for request 1
Request 2: Correlation ID : 6
cc.cache overflow, so correlation ID 5 gets overwritten with 6
cc.Delete now goes and deletes 6 instead of 5, since it only checks for correlation ID and not request.


Should be make the looking up and deleting a single function protected by a lock?

Let me know if this scenario cannot happen.

Reply:
I don't follow regarding the race. `Delete()` was called on the correlation ID of the modified request. I've changed the code to get rid of `Delete()`, this might make it easier to understand.